File: Try_GameEngine5/src/com/example/try_gameengine/test/EnemyFactory.java
package com.example.try_gameengine.test;

import java.lang.reflect.InvocationTargetException;
import java.util.List;

import android.graphics.Bitmap;

import com.example.try_gameengine.action.MovementAction;
import com.example.try_gameengine.action.MovementActionFactory;
import com.example.try_gameengine.action.MovementActionInfo;
import com.example.try_gameengine.action.MovementActionSetWithOutThread;
import com.example.try_gameengine.action.MovementActionSetWithThread;
import com.example.try_gameengine.action.MovementDecorator;

public class EnemyFactory {
	private static Bitmap redEnemyBitmap;
	private static Bitmap blueEnemyBitmap;
	
	public Enemy createRedEnemy() {
		return new RedEnemy(0, 0);
	}

	public Enemy createBlueEnemy() {
		return new BlueEnemy(10, 10);
	}

	public Enemy createRedEnemy(int[] enemyInfo){
		return new RedEnemy(enemyInfo[0], enemyInfo[1]);
	}

	public Enemy createBlueEnemy(int[] enemyInfo) {
		return new BlueEnemy(10, 10);
	}
	
	public Enemy createRLRedEnemy(int[] enemyInfo){
		return new RedEnemy(enemyInfo[0], enemyInfo[1], new RLMovementActionFactory().createMovementAction().initMovementAction());
	}
	
	public Enemy createLRRedEnemy(int[] enemyInfo){
		return new RedEnemy(enemyInfo[0], enemyInfo[1], new LRMovementActionFactory().createMovementAction().initMovementAction());
	}
	
	public Enemy createRLBlueEnemy(int[] enemyInfo){
		return new BlueEnemy(enemyInfo[0], enemyInfo[1], new RLMovementActionFactory().createMovementAction().initMovementAction());
	}
	
	public Enemy createLRBlueEnemy(int[] enemyInfo){
		return new BlueEnemy(enemyInfo[0], enemyInfo[1], new LRMovementActionFactory().createMovementAction().initMovementAction());
	}
	
	public Enemy createSpecialEnemy(Class<? extends Enemy> enemyClass, Class<? extends MovementActionFactory> actionFactoryClass,int[] enemyInfo){
		Enemy enemy = null;
		MovementAction action = null;
		try {
			if(actionFactoryClass!=null)
				action = actionFactoryClass.newInstance().createMovementAction().initMovementAction();
			enemy = enemyClass.getConstructor(int.class, int.class, MovementAction.class).newInstance(enemyInfo[0], enemyInfo[1], action);
		} catch (InstantiationException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		} catch (IllegalAccessException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		} catch (IllegalArgumentException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		} catch (InvocationTargetException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		} catch (NoSuchMethodException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}
		return enemy;
	}
	
	public Enemy createSpecialEnemy2(Class<? extends Enemy> enemyClass, Class<? extends MovementActionFactory> actionFactoryClass,int[] enemyInfo, List<MovementActionInfo> infos){
		Enemy enemy = null;
		MovementAction action = null;
		try {
			if(actionFactoryClass!=null)
				action = actionFactoryClass.newInstance().createMovementAction(infos).initMovementAction();
			enemy = enemyClass.getConstructor(int.class, int.class, MovementAction.class).newInstance(enemyInfo[0], enemyInfo[1], action);
		} catch (InstantiationException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		} catch (IllegalAccessException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		} catch (IllegalArgumentException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		} catch (InvocationTargetException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		} catch (NoSuchMethodException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}
		return enemy;
	}
	
	public Enemy createSpecialEnemy3(Class<? extends Enemy> enemyClass, Class<? extends MovementActionFactory> actionFactoryClass,int[] enemyInfo, List<MovementActionInfo> infos, List<Class<? extends MovementDecorator>> decoratorClassList){
		Enemy enemy = null;
		MovementAction action = null;
		try {
			if(actionFactoryClass!=null)
				action = actionFactoryClass.newInstance().createMovementAction(infos, decoratorClassList);
			MovementAction set = new MovementActionSetWithOutThread();
			enemy = enemyClass.getConstructor(int.class, int.class, MovementAction.class).newInstance(enemyInfo[0], enemyInfo[1], set.addMovementAction(action).initMovementAction());
		} catch (InstantiationException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		} catch (IllegalAccessException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		} catch (IllegalArgumentException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		} catch (InvocationTargetException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		} catch (NoSuchMethodException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}
		return enemy;
	}
	
	public Enemy createSpecialEnemy4(Class<? extends Enemy> enemyClass, Class<? extends MovementActionFactory> actionFactoryClass,int[] enemyInfo, List<Class<? extends MovementDecorator>> decoratorClassList){
		Enemy enemy = null;
		MovementAction action = null;
		try {
			if(actionFactoryClass!=null)
				action = actionFactoryClass.newInstance().createMovementActionByDecorator(decoratorClassList);
			MovementAction set = new MovementActionSetWithOutThread();
			enemy = enemyClass.getConstructor(int.class, int.class, MovementAction.class).newInstance(enemyInfo[0], enemyInfo[1], set.addMovementAction(action).initMovementAction());
		} catch (InstantiationException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		} catch (IllegalAccessException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		} catch (IllegalArgumentException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		} catch (InvocationTargetException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		} catch (NoSuchMethodException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}
		return enemy;
	}
	
	public Enemy createSpecialEnemy5(Class<? extends Enemy> enemyClass, int[] enemyInfo, MovementAction action){
		Enemy enemy = null;
		try {
			enemy = enemyClass.getConstructor(int.class, int.class, MovementAction.class).newInstance(enemyInfo[0], enemyInfo[1], new MovementActionSetWithThread().addMovementAction(action).initMovementAction());
		} catch (InstantiationException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		} catch (IllegalAccessException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		} catch (IllegalArgumentException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		} catch (InvocationTargetException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		} catch (NoSuchMethodException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}
		return enemy;
	}
	
	public static void setRedEnemyBitmap(Bitmap redEnemyBitmap){
		EnemyFactory.redEnemyBitmap = redEnemyBitmap;
	}
	
	public static void setBlueEnemyBitmap(Bitmap blueEnemyBitmap){
		EnemyFactory.blueEnemyBitmap = blueEnemyBitmap;
	}
	
	public static Bitmap getRedEnemyBitmap(){
		return EnemyFactory.redEnemyBitmap;
	}
	
	public static Bitmap getBlueEnemyBitmap(){
		return EnemyFactory.blueEnemyBitmap;
	}
}

File: Try_GameEngine5/src/com/example/try_gameengine/test/EnemyFactory.kt
package com.example.try_gameengine.test

import android.graphics.Bitmap
import com.example.try_gameengine.action.MovementAction
import com.example.try_gameengine.action.MovementActionFactory
import com.example.try_gameengine.action.MovementActionInfo
import com.example.try_gameengine.action.MovementActionSetWithOutThread
import com.example.try_gameengine.action.MovementActionSetWithThread
import com.example.try_gameengine.action.MovementDecorator
import java.lang.reflect.InvocationTargetException

class EnemyFactory {
    fun createRedEnemy(): Enemy {
        return RedEnemy(0, 0)
    }

    fun createBlueEnemy(): Enemy {
        return BlueEnemy(10, 10)
    }

    fun createRedEnemy(enemyInfo: IntArray): Enemy {
        return RedEnemy(enemyInfo[0], enemyInfo[1])
    }

    fun createBlueEnemy(enemyInfo: IntArray?): Enemy {
        return BlueEnemy(10, 10)
    }

    fun createRLRedEnemy(enemyInfo: IntArray): Enemy {
        return RedEnemy(
            enemyInfo[0],
            enemyInfo[1],
            RLMovementActionFactory().createMovementAction().initMovementAction()
        )
    }

    fun createLRRedEnemy(enemyInfo: IntArray): Enemy {
        return RedEnemy(
            enemyInfo[0],
            enemyInfo[1],
            LRMovementActionFactory().createMovementAction().initMovementAction()
        )
    }

    fun createRLBlueEnemy(enemyInfo: IntArray): Enemy {
        return BlueEnemy(
            enemyInfo[0],
            enemyInfo[1],
            RLMovementActionFactory().createMovementAction().initMovementAction()
        )
    }

    fun createLRBlueEnemy(enemyInfo: IntArray): Enemy {
        return BlueEnemy(
            enemyInfo[0],
            enemyInfo[1],
            LRMovementActionFactory().createMovementAction().initMovementAction()
        )
    }

    fun createSpecialEnemy(
        enemyClass: Class<out Enemy?>,
        actionFactoryClass: Class<out MovementActionFactory?>?,
        enemyInfo: IntArray
    ): Enemy? {
        var enemy: Enemy? = null
        var action: MovementAction? = null
        try {
            if (actionFactoryClass != null) action =
                actionFactoryClass.newInstance().createMovementAction().initMovementAction()
            enemy = enemyClass.getConstructor(
                Int::class.javaPrimitiveType,
                Int::class.javaPrimitiveType,
                MovementAction::class.java
            ).newInstance(enemyInfo[0], enemyInfo[1], action)
        } catch (e: InstantiationException) {
            // TODO Auto-generated catch block
            e.printStackTrace()
        } catch (e: IllegalAccessException) {
            // TODO Auto-generated catch block
            e.printStackTrace()
        } catch (e: IllegalArgumentException) {
            // TODO Auto-generated catch block
            e.printStackTrace()
        } catch (e: InvocationTargetException) {
            // TODO Auto-generated catch block
            e.printStackTrace()
        } catch (e: NoSuchMethodException) {
            // TODO Auto-generated catch block
            e.printStackTrace()
        }
        return enemy
    }

    fun createSpecialEnemy2(
        enemyClass: Class<out Enemy?>,
        actionFactoryClass: Class<out MovementActionFactory?>?,
        enemyInfo: IntArray,
        infos: MutableList<MovementActionInfo?>?
    ): Enemy? {
        var enemy: Enemy? = null
        var action: MovementAction? = null
        try {
            if (actionFactoryClass != null) action =
                actionFactoryClass.newInstance().createMovementAction(infos).initMovementAction()
            enemy = enemyClass.getConstructor(
                Int::class.javaPrimitiveType,
                Int::class.javaPrimitiveType,
                MovementAction::class.java
            ).newInstance(enemyInfo[0], enemyInfo[1], action)
        } catch (e: InstantiationException) {
            // TODO Auto-generated catch block
            e.printStackTrace()
        } catch (e: IllegalAccessException) {
            // TODO Auto-generated catch block
            e.printStackTrace()
        } catch (e: IllegalArgumentException) {
            // TODO Auto-generated catch block
            e.printStackTrace()
        } catch (e: InvocationTargetException) {
            // TODO Auto-generated catch block
            e.printStackTrace()
        } catch (e: NoSuchMethodException) {
            // TODO Auto-generated catch block
            e.printStackTrace()
        }
        return enemy
    }

    fun createSpecialEnemy3(
        enemyClass: Class<out Enemy?>,
        actionFactoryClass: Class<out MovementActionFactory?>?,
        enemyInfo: IntArray,
        infos: MutableList<MovementActionInfo?>?,
        decoratorClassList: MutableList<Class<out MovementDecorator?>?>?
    ): Enemy? {
        var enemy: Enemy? = null
        var action: MovementAction? = null
        try {
            if (actionFactoryClass != null) action =
                actionFactoryClass.newInstance().createMovementAction(infos, decoratorClassList)
            val set: MovementAction = MovementActionSetWithOutThread()
            enemy = enemyClass.getConstructor(
                Int::class.javaPrimitiveType,
                Int::class.javaPrimitiveType,
                MovementAction::class.java
            ).newInstance(
                enemyInfo[0],
                enemyInfo[1],
                set.addMovementAction(action).initMovementAction()
            )
        } catch (e: InstantiationException) {
            // TODO Auto-generated catch block
            e.printStackTrace()
        } catch (e: IllegalAccessException) {
            // TODO Auto-generated catch block
            e.printStackTrace()
        } catch (e: IllegalArgumentException) {
            // TODO Auto-generated catch block
            e.printStackTrace()
        } catch (e: InvocationTargetException) {
            // TODO Auto-generated catch block
            e.printStackTrace()
        } catch (e: NoSuchMethodException) {
            // TODO Auto-generated catch block
            e.printStackTrace()
        }
        return enemy
    }

    fun createSpecialEnemy4(
        enemyClass: Class<out Enemy?>,
        actionFactoryClass: Class<out MovementActionFactory?>?,
        enemyInfo: IntArray,
        decoratorClassList: MutableList<Class<out MovementDecorator?>?>
    ): Enemy? {
        var enemy: Enemy? = null
        var action: MovementAction? = null
        try {
            if (actionFactoryClass != null) action =
                actionFactoryClass.newInstance().createMovementActionByDecorator(decoratorClassList)
            val set: MovementAction = MovementActionSetWithOutThread()
            enemy = enemyClass.getConstructor(
                Int::class.javaPrimitiveType,
                Int::class.javaPrimitiveType,
                MovementAction::class.java
            ).newInstance(
                enemyInfo[0],
                enemyInfo[1],
                set.addMovementAction(action).initMovementAction()
            )
        } catch (e: InstantiationException) {
            // TODO Auto-generated catch block
            e.printStackTrace()
        } catch (e: IllegalAccessException) {
            // TODO Auto-generated catch block
            e.printStackTrace()
        } catch (e: IllegalArgumentException) {
            // TODO Auto-generated catch block
            e.printStackTrace()
        } catch (e: InvocationTargetException) {
            // TODO Auto-generated catch block
            e.printStackTrace()
        } catch (e: NoSuchMethodException) {
            // TODO Auto-generated catch block
            e.printStackTrace()
        }
        return enemy
    }

    fun createSpecialEnemy5(
        enemyClass: Class<out Enemy?>,
        enemyInfo: IntArray,
        action: MovementAction?
    ): Enemy? {
        var enemy: Enemy? = null
        try {
            enemy = enemyClass.getConstructor(
                Int::class.javaPrimitiveType,
                Int::class.javaPrimitiveType,
                MovementAction::class.java
            ).newInstance(
                enemyInfo[0],
                enemyInfo[1],
                MovementActionSetWithThread().addMovementAction(action).initMovementAction()
            )
        } catch (e: InstantiationException) {
            // TODO Auto-generated catch block
            e.printStackTrace()
        } catch (e: IllegalAccessException) {
            // TODO Auto-generated catch block
            e.printStackTrace()
        } catch (e: IllegalArgumentException) {
            // TODO Auto-generated catch block
            e.printStackTrace()
        } catch (e: InvocationTargetException) {
            // TODO Auto-generated catch block
            e.printStackTrace()
        } catch (e: NoSuchMethodException) {
            // TODO Auto-generated catch block
            e.printStackTrace()
        }
        return enemy
    }

    companion object {
        var redEnemyBitmap: Bitmap? = null
        var blueEnemyBitmap: Bitmap? = null
    }
}
